File: sdk/verifiers/rust/src/canonical.rs
// Copyright 2026 Pipelock contributors
// SPDX-License-Identifier: Apache-2.0

use serde_json::{Map, Number, Value};
use unicode_normalization::UnicodeNormalization;

#[derive(Clone, Copy)]
enum NestedKind {
    ActionRecord,
    Redaction,
    Shield,
    TaintSource,
    KeyTransition,
    SessionControl,
    SessionOpen,
    SessionHeartbeat,
    SessionClose,
}

#[derive(Clone, Copy)]
struct FieldSpec {
    name: &'static str,
    omitempty: bool,
    nested: Option<NestedKind>,
}

const ACTION_RECORD_FIELDS: &[FieldSpec] = &[
    field("version", false),
    field("action_id", false),
    field("parent_action_id", true),
    field("action_type", false),
    field("timestamp", false),
    field("principal", false),
    field("actor", false),
    field("delegation_chain", false),
    field("target", false),
    field("intent", true),
    field("data_classes_in", true),
    field("data_classes_out", true),
    field("side_effect_class", false),
    field("reversibility", false),
    field("policy_hash", false),
    field("verdict", false),
    field("decision_phase", true),
    field("defer_id", true),
    field("resolution_policy", true),
    field("resolution_source", true),
    field("session_id", true),
    field("session_id_original", true),
    field("session_taint_level", true),
    field("session_contaminated", true),
    nested_field("recent_taint_sources", true, NestedKind::TaintSource),
    field("session_task_id", true),
    field("session_task_label", true),
    field("authority_kind", true),
    field("taint_decision", true),
    field("taint_decision_reason", true),
    field("task_override_applied", true),
    field("contract_winning_source", true),
    field("contract_live_verdict", true),
    field("contract_policy_sources", true),
    field("contract_rule_id", true),
    field("active_manifest_hash", true),
    field("contract_hash", true),
    field("contract_selector_id", true),
    field("contract_generation", true),
    field("transport", false),
    field("method", true),
    field("layer", true),
    field("pattern", true),
    field("severity", true),
    nested_field("redaction", true, NestedKind::Redaction),
    nested_field("shield", true, NestedKind::Shield),
    field("request_id", true),
    field("chain_prev_hash", false),
    field("chain_seq", false),
    field("run_nonce", true),
    nested_field("key_transition", true, NestedKind::KeyTransition),
    nested_field("session_control", true, NestedKind::SessionControl),
    field("venue", true),
    field("jurisdiction", true),
    field("rulebook_id", true),
    field("remedy_class", true),
    field("contestation_window", true),
    field("precedent_refs", true),
];

const RECEIPT_FIELDS: &[FieldSpec] = &[
    field("version", false),
    nested_field("action_record", false, NestedKind::ActionRecord),
    field("signature", false),
    field("signer_key", false),
];

const REDACTION_FIELDS: &[FieldSpec] = &[
    field("profile", true),
    field("provider", true),
    field("parser", true),
    field("total_redactions", true),
    field("by_class", true),
    field("cache_boundary_kept", true),
];

// SHIELD_FIELDS mirrors receipt.ShieldSummary in Go struct-declaration order.
// A shield-bearing receipt must reorder the nested shield object to this exact
// order or it recomputes a different signing hash than the Go signer produced.
const SHIELD_FIELDS: &[FieldSpec] = &[
    field("pipeline", true),
    field("total_rewrites", true),
    field("extension_probes", true),
    field("tracking_beacons", true),
    field("agent_traps", true),
    field("fingerprint_shim_injected", true),
    field("svg_foreign_objects", true),
    field("svg_event_handlers", true),
    field("svg_external_references", true),
    field("svg_hidden_text", true),
    field("svg_animation_injections", true),
    field("body_bytes", true),
    field("scanned_bytes", true),
    field("partial", true),
    field("adaptive_signals_recorded", true),
    field("adaptive_signal_max_per_body", true),
];

const TAINT_SOURCE_FIELDS: &[FieldSpec] = &[
    field("url", false),
    field("kind", false),
    field("level", false),
    field("timestamp", false),
    field("receipt_id", true),
    field("match_reason", true),
];

// KEY_TRANSITION_FIELDS mirrors receipt.KeyTransition in Go struct-declaration
// order. Stamped on a segment-genesis receipt after a signing-key rotation; the
// nested object MUST be reordered to this exact order or a rotated-segment
// receipt recomputes a different signing hash than the Go signer produced.
const KEY_TRANSITION_FIELDS: &[FieldSpec] = &[
    field("prior_signer_key", false),
    field("prior_chain_seq", false),
    field("prior_chain_hash", false),
];

const SESSION_CONTROL_FIELDS: &[FieldSpec] = &[
    field("kind", false),
    nested_field("open", true, NestedKind::SessionOpen),
    nested_field("heartbeat", true, NestedKind::SessionHeartbeat),
    nested_field("close", true, NestedKind::SessionClose),
];

const SESSION_OPEN_FIELDS: &[FieldSpec] = &[
    field("run_nonce", false),
    field("open_nonce", false),
    field("recorder_session", false),
    field("policy_hash", false),
    field("signer_key_epoch", false),
    field("heartbeat_seconds", false),
    field("chain_open_seq", false),
    field("prior_chain_head", true),
    field("prior_chain_seq", true),
    field("genesis_hash", true),
    field("genesis_anchor_head", true),
    field("genesis_anchor_log", true),
    field("posture_capsule_sha256", true),
    field("posture_signer_key_id", true),
    field("containment_nonce", true),
    field("contained_uid", true),
];

const SESSION_HEARTBEAT_FIELDS: &[FieldSpec] = &[
    field("run_nonce", false),
    field("open_nonce", false),
    field("beat", false),
    field("chain_head", false),
    field("chain_seq_head", false),
    field("heartbeat_time", false),
    field("fsync_errors_gated", false),
    field("durability_blocks", false),
];

const SESSION_CLOSE_FIELDS: &[FieldSpec] = &[
    field("run_nonce", false),
    field("open_nonce", false),
    field("final_seq", false),
    field("root_hash", false),
    field("receipt_count", false),
    field("close_reason", false),
    field("fsync_errors_gated", false),
    field("durability_blocks", false),
];

const fn field(name: &'static str, omitempty: bool) -> FieldSpec {
    FieldSpec {
        name,
        omitempty,
        nested: None,
    }
}

const fn nested_field(name: &'static str, omitempty: bool, nested: NestedKind) -> FieldSpec {
    FieldSpec {
        name,
        omitempty,
        nested: Some(nested),
    }
}

pub fn canonicalize_action_record(action_record: &Value) -> Vec<u8> {
    canonical_json_bytes(&order_struct(action_record, ACTION_RECORD_FIELDS))
}

pub fn canonicalize_receipt(receipt: &Value) -> Vec<u8> {
    canonical_json_bytes(&order_struct(receipt, RECEIPT_FIELDS))
}

pub fn canonical_json_string(value: &Value) -> String {
    go_html_escape(&serde_json::to_string(value).expect("serialize JSON value"))
}

pub fn canonicalize_jcs_value(value: &Value) -> Result<Vec<u8>, String> {
    let mut out = String::new();
    canonicalize_jcs_into(&mut out, value)?;
    Ok(out.into_bytes())
}

fn canonical_json_bytes(value: &Value) -> Vec<u8> {
    canonical_json_string(value).into_bytes()
}

fn order_struct(value: &Value, fields: &[FieldSpec]) -> Value {
    let input = value.as_object();
    let mut out = Map::new();
    for spec in fields {
        let mut field_value = input.and_then(|object| object.get(spec.name)).cloned();
        if field_value.is_none() {
            if spec.omitempty {
                continue;
            }
            field_value = Some(zero_value(spec.name, spec.nested));
        }
        let mut field_value = field_value.expect("field value set");
        field_value = match spec.nested {
            Some(NestedKind::ActionRecord) if field_value.is_object() => {
                order_struct(&field_value, ACTION_RECORD_FIELDS)
            }
            Some(NestedKind::Redaction) if field_value.is_object() => {
                order_struct(&field_value, REDACTION_FIELDS)
            }
            Some(NestedKind::Shield) if field_value.is_object() => {
                order_struct(&field_value, SHIELD_FIELDS)
            }
            Some(NestedKind::TaintSource) if field_value.is_array() => Value::Array(
                field_value
                    .as_array()
                    .expect("checked array")
                    .iter()
                    .map(|item| {
                        if item.is_object() {
                            order_struct(item, TAINT_SOURCE_FIELDS)
                        } else {
                            item.clone()
                        }
                    })
                    .collect(),
            ),
            Some(NestedKind::KeyTransition) if field_value.is_object() => {
                order_struct(&field_value, KEY_TRANSITION_FIELDS)
            }
            Some(NestedKind::SessionControl) if field_value.is_object() => {
                order_struct(&field_value, SESSION_CONTROL_FIELDS)
            }
            Some(NestedKind::SessionOpen) if field_value.is_object() => {
                order_struct(&field_value, SESSION_OPEN_FIELDS)
            }
            Some(NestedKind::SessionHeartbeat) if field_value.is_object() => {
                order_struct(&field_value, SESSION_HEARTBEAT_FIELDS)
            }
            Some(NestedKind::SessionClose) if field_value.is_object() => {
                order_struct(&field_value, SESSION_CLOSE_FIELDS)
            }
            _ => normalize_maps(&field_value),
        };
        if spec.omitempty && is_go_zero(&field_value) {
            continue;
        }
        out.insert(spec.name.to_string(), field_value);
    }
    Value::Object(out)
}

fn zero_value(name: &str, nested: Option<NestedKind>) -> Value {
    if matches!(nested, Some(NestedKind::ActionRecord)) {
        return Value::Object(Map::new());
    }
    match name {
        "version" | "chain_seq" | "level" | "prior_chain_seq" | "heartbeat_seconds"
        | "chain_open_seq" | "beat" | "chain_seq_head" | "fsync_errors_gated"
        | "durability_blocks" | "final_seq" | "receipt_count" => Value::Number(Number::from(0)),
        "delegation_chain" => Value::Null,
        "timestamp" => Value::String("0001-01-01T00:00:00Z".to_string()),
        _ => Value::String(String::new()),
    }
}

fn is_go_zero(value: &Value) -> bool {
    match value {
        Value::Null => true,
        Value::Bool(value) => !*value,
        Value::Number(value) => value.as_i64() == Some(0) || value.as_u64() == Some(0),
        Value::String(value) => value.is_empty(),
        Value::Array(value) => value.is_empty(),
        Value::Object(value) => value.is_empty(),
    }
}

fn normalize_maps(value: &Value) -> Value {
    match value {
        Value::Array(items) => Value::Array(items.iter().map(normalize_maps).collect()),
        Value::Object(object) => {
            let mut keys = object.keys().collect::<Vec<_>>();
            keys.sort();
            let mut out = Map::new();
            for key in keys {
                out.insert(key.clone(), normalize_maps(&object[key]));
            }
            Value::Object(out)
        }
        _ => value.clone(),
    }
}

fn canonicalize_jcs_into(out: &mut String, value: &Value) -> Result<(), String> {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(true) => out.push_str("true"),
        Value::Bool(false) => out.push_str("false"),
        Value::Number(number) => {
            if number.is_i64() || number.is_u64() {
                out.push_str(&number.to_string());
            } else {
                return Err(format!("float not allowed in canonicalization: {number}"));
            }
        }
        Value::String(value) => out.push_str(&encode_jcs_string(value)),
        Value::Array(items) => {
            out.push('[');
            for (index, item) in items.iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                canonicalize_jcs_into(out, item)?;
            }
            out.push(']');
        }
        Value::Object(object) => {
            let mut pairs: Vec<(String, &Value)> = object
                .iter()
                .map(|(key, value)| (key.nfc().collect::<String>(), value))
                .collect();
            pairs.sort_by(|a, b| a.0.cmp(&b.0));
            for index in 1..pairs.len() {
                if pairs[index].0 == pairs[index - 1].0 {
                    return Err(format!("NFC collision on key {:?}", pairs[index].0));
                }
            }
            out.push('{');
            for (index, (key, value)) in pairs.iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                out.push_str(&encode_jcs_string(key));
                out.push(':');
                canonicalize_jcs_into(out, value)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

fn encode_jcs_string(s: &str) -> String {
    let normalized: String = s.nfc().collect();
    let mut out = String::with_capacity(normalized.len() + 2);
    out.push('"');
    for ch in normalized.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0c}' => out.push_str("\\f"),
            '<' => out.push_str("\\u003c"),
            '>' => out.push_str("\\u003e"),
            '&' => out.push_str("\\u0026"),
            '\u{2028}' => out.push_str("\\u2028"),
            '\u{2029}' => out.push_str("\\u2029"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

fn go_html_escape(serialized: &str) -> String {
    let mut out = String::with_capacity(serialized.len());
    for ch in serialized.chars() {
        match ch {
            '<' => out.push_str("\\u003c"),
            '>' => out.push_str("\\u003e"),
            '&' => out.push_str("\\u0026"),
            '\u{2028}' => out.push_str("\\u2028"),
            '\u{2029}' => out.push_str("\\u2029"),
            _ => out.push(ch),
        }
    }
    out
}

File: sdk/verifiers/ts/src/canonical.ts
import type { ActionRecord, Receipt, JSONValue } from "./types.js";

type FieldSpec = readonly [name: string, omitempty: boolean, nested?: NestedKind];
type NestedKind =
  | "action_record"
  | "redaction"
  | "shield"
  | "taint_source"
  | "key_transition"
  | "session_control"
  | "session_open"
  | "session_heartbeat"
  | "session_close";

const actionRecordFields: readonly FieldSpec[] = [
  ["version", false],
  ["action_id", false],
  ["parent_action_id", true],
  ["action_type", false],
  ["timestamp", false],
  ["principal", false],
  ["actor", false],
  ["delegation_chain", false],
  ["target", false],
  ["intent", true],
  ["data_classes_in", true],
  ["data_classes_out", true],
  ["side_effect_class", false],
  ["reversibility", false],
  ["policy_hash", false],
  ["verdict", false],
  ["decision_phase", true],
  ["defer_id", true],
  ["resolution_policy", true],
  ["resolution_source", true],
  ["session_id", true],
  ["session_id_original", true],
  ["session_taint_level", true],
  ["session_contaminated", true],
  ["recent_taint_sources", true, "taint_source"],
  ["session_task_id", true],
  ["session_task_label", true],
  ["authority_kind", true],
  ["taint_decision", true],
  ["taint_decision_reason", true],
  ["task_override_applied", true],
  ["contract_winning_source", true],
  ["contract_live_verdict", true],
  ["contract_policy_sources", true],
  ["contract_rule_id", true],
  ["active_manifest_hash", true],
  ["contract_hash", true],
  ["contract_selector_id", true],
  ["contract_generation", true],
  ["transport", false],
  ["method", true],
  ["layer", true],
  ["pattern", true],
  ["severity", true],
  ["redaction", true, "redaction"],
  ["shield", true, "shield"],
  ["request_id", true],
  ["chain_prev_hash", false],
  ["chain_seq", false],
  ["run_nonce", true],
  ["key_transition", true, "key_transition"],
  ["session_control", true, "session_control"],
  ["venue", true],
  ["jurisdiction", true],
  ["rulebook_id", true],
  ["remedy_class", true],
  ["contestation_window", true],
  ["precedent_refs", true],
];

const receiptFields: readonly FieldSpec[] = [
  ["version", false],
  ["action_record", false, "action_record"],
  ["signature", false],
  ["signer_key", false],
];

const redactionFields: readonly FieldSpec[] = [
  ["profile", true],
  ["provider", true],
  ["parser", true],
  ["total_redactions", true],
  ["by_class", true],
  ["cache_boundary_kept", true],
];

// shieldFields mirrors receipt.ShieldSummary in Go struct-declaration order.
// The nested shield object MUST be reordered to this exact order before
// serialization or a shield-bearing receipt recomputes a different signing
// hash than the Go signer produced.
const shieldFields: readonly FieldSpec[] = [
  ["pipeline", true],
  ["total_rewrites", true],
  ["extension_probes", true],
  ["tracking_beacons", true],
  ["agent_traps", true],
  ["fingerprint_shim_injected", true],
  ["svg_foreign_objects", true],
  ["svg_event_handlers", true],
  ["svg_external_references", true],
  ["svg_hidden_text", true],
  ["svg_animation_injections", true],
  ["body_bytes", true],
  ["scanned_bytes", true],
  ["partial", true],
  ["adaptive_signals_recorded", true],
  ["adaptive_signal_max_per_body", true],
];

const taintSourceFields: readonly FieldSpec[] = [
  ["url", false],
  ["kind", false],
  ["level", false],
  ["timestamp", false],
  ["receipt_id", true],
  ["match_reason", true],
];

// keyTransitionFields mirrors receipt.KeyTransition in Go struct-declaration
// order. Stamped on a segment-genesis receipt after a signing-key rotation; the
// nested object MUST be reordered to this exact order before serialization or a
// rotated-segment receipt recomputes a different signing hash than the Go signer
// produced. All three fields are required (no omitempty) - the marker is only
// present as a whole when a rotation occurred.
const keyTransitionFields: readonly FieldSpec[] = [
  ["prior_signer_key", false],
  ["prior_chain_seq", false],
  ["prior_chain_hash", false],
];

const sessionControlFields: readonly FieldSpec[] = [
  ["kind", false],
  ["open", true, "session_open"],
  ["heartbeat", true, "session_heartbeat"],
  ["close", true, "session_close"],
];

const sessionOpenFields: readonly FieldSpec[] = [
  ["run_nonce", false],
  ["open_nonce", false],
  ["recorder_session", false],
  ["policy_hash", false],
  ["signer_key_epoch", false],
  ["heartbeat_seconds", false],
  ["chain_open_seq", false],
  ["prior_chain_head", true],
  ["prior_chain_seq", true],
  ["genesis_hash", true],
  ["genesis_anchor_head", true],
  ["genesis_anchor_log", true],
  ["posture_capsule_sha256", true],
  ["posture_signer_key_id", true],
  ["containment_nonce", true],
  ["contained_uid", true],
];

const sessionHeartbeatFields: readonly FieldSpec[] = [
  ["run_nonce", false],
  ["open_nonce", false],
  ["beat", false],
  ["chain_head", false],
  ["chain_seq_head", false],
  ["heartbeat_time", false],
  ["fsync_errors_gated", false],
  ["durability_blocks", false],
];

const sessionCloseFields: readonly FieldSpec[] = [
  ["run_nonce", false],
  ["open_nonce", false],
  ["final_seq", false],
  ["root_hash", false],
  ["receipt_count", false],
  ["close_reason", false],
  ["fsync_errors_gated", false],
  ["durability_blocks", false],
];

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isGoZero(value: unknown): boolean {
  if (value === null || value === undefined) return true;
  if (typeof value === "boolean") return !value;
  if (typeof value === "number") return value === 0;
  if (typeof value === "string") return value === "";
  if (Array.isArray(value)) return value.length === 0;
  if (isPlainObject(value)) return Object.keys(value).length === 0;
  return false;
}

function orderStruct(
  value: Record<string, unknown>,
  fields: readonly FieldSpec[],
): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [name, omitempty, nested] of fields) {
    let fieldValue = value[name];
    if (!Object.prototype.hasOwnProperty.call(value, name)) {
      if (omitempty) continue;
      fieldValue = zeroValue(name, nested);
    }
    if (nested === "action_record" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, actionRecordFields);
    } else if (nested === "redaction" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, redactionFields);
    } else if (nested === "shield" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, shieldFields);
    } else if (nested === "taint_source" && Array.isArray(fieldValue)) {
      fieldValue = fieldValue.map((item) =>
        isPlainObject(item) ? orderStruct(item, taintSourceFields) : item,
      );
    } else if (nested === "key_transition" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, keyTransitionFields);
    } else if (nested === "session_control" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, sessionControlFields);
    } else if (nested === "session_open" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, sessionOpenFields);
    } else if (nested === "session_heartbeat" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, sessionHeartbeatFields);
    } else if (nested === "session_close" && isPlainObject(fieldValue)) {
      fieldValue = orderStruct(fieldValue, sessionCloseFields);
    } else {
      fieldValue = normalizeMaps(fieldValue);
    }
    if (omitempty && isGoZero(fieldValue)) continue;
    out[name] = fieldValue;
  }
  return out;
}

function zeroValue(name: string, nested?: NestedKind): unknown {
  if (nested === "action_record") return {};
  if (
    name === "version" ||
    name === "chain_seq" ||
    name === "level" ||
    name === "prior_chain_seq" ||
    name === "heartbeat_seconds" ||
    name === "chain_open_seq" ||
    name === "beat" ||
    name === "chain_seq_head" ||
    name === "fsync_errors_gated" ||
    name === "durability_blocks" ||
    name === "final_seq" ||
    name === "receipt_count"
  )
    return 0;
  if (name === "delegation_chain") return null;
  if (name === "timestamp") return "0001-01-01T00:00:00Z";
  return "";
}

function normalizeMaps(value: unknown): unknown {
  if (Array.isArray(value)) return value.map((item) => normalizeMaps(item));
  if (!isPlainObject(value)) return value;
  const out: Record<string, unknown> = {};
  for (const key of Object.keys(value).sort(compareCodePointStrings)) {
    const item = value[key];
    if (item === undefined) continue;
    out[key] = normalizeMaps(item);
  }
  return out;
}

function compareCodePointStrings(a: string, b: string): number {
  const left = Array.from(a);
  const right = Array.from(b);
  const n = Math.min(left.length, right.length);
  for (let i = 0; i < n; i++) {
    const leftCodePoint = left[i]?.codePointAt(0) ?? 0;
    const rightCodePoint = right[i]?.codePointAt(0) ?? 0;
    if (leftCodePoint !== rightCodePoint) return leftCodePoint - rightCodePoint;
  }
  return left.length - right.length;
}

function stringifyCompact(value: unknown): string {
  return JSON.stringify(value);
}

function goHTMLEscape(serialized: string): string {
  return serialized
    .replace(/</g, "\\u003c")
    .replace(/>/g, "\\u003e")
    .replace(/&/g, "\\u0026")
    .replace(/\u2028/g, "\\u2028")
    .replace(/\u2029/g, "\\u2029");
}

export function canonicalizeActionRecord(actionRecord: ActionRecord): Buffer {
  return Buffer.from(
    goHTMLEscape(stringifyCompact(orderStruct(actionRecord, actionRecordFields))),
    "utf8",
  );
}

export function canonicalizeReceipt(receipt: Receipt): Buffer {
  return Buffer.from(goHTMLEscape(stringifyCompact(orderStruct(receipt, receiptFields))), "utf8");
}

export function canonicalJSONString(value: JSONValue): string {
  return goHTMLEscape(stringifyCompact(value));
}
